cranelift/codegen/src/isa/riscv64/abi.rs

```diff
@@ -111,11 +111,18 @@ impl ABIMachineSpec for Riscv64MachineDeps {
         // Stack space.
         let mut next_stack: u32 = 0;
 
-        // In the SystemV ABI, the return area pointer is the first argument,
-        // so we need to leave room for it if required.
-        if add_ret_area_ptr && call_conv != CallConv::Tail && call_conv != CallConv::Winch {
+        let ret_area_ptr = if add_ret_area_ptr {
+            assert!(ArgsOrRets::Args == args_or_rets);
             next_x_reg += 1;
-        }
+            Some(ABIArg::reg(
+                x_reg(x_start).to_real_reg().unwrap(),
+                I64,
+                ir::ArgumentExtension::None,
+                ir::ArgumentPurpose::Normal,
+            ))
```

Comment:
Feel similarly here

Reply:
This really needs to be the first argument register. If next_x_reg - 1 is not x_start at this point, that is a bug.